dbs-virtio-devices: add vsock device.

namespaceXP opened the pull request:
Add virtio vsock device.

namespaceXP committed:
dbs-virtio-devices: add epoll_handler for vsock.

Add VsockEpollHandler, which implements the runtime logic of vsock
device.

Signed-off-by: Beiyue <beiyue@linux.alibaba.com>

namespaceXP committed:
dbs-virtio-devices: add Vsock device.

Add virtio vsock device.

Signed-off-by: Beiyue <beiyue@linux.alibaba.com>

namespaceXP committed:
dbs-virtio-devices: add unit tests for vsock.

Add some unit test cases for vsock.

Signed-off-by: Beiyue <beiyue@linux.alibaba.com>

wllenyj commented on crates/dbs-virtio-devices/src/epoll_handler.rs:
```diff
@@ -0,0 +1,310 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Portions Copyright (C) 2019 Alibaba Cloud Computing. All rights reserved.
```
Unified Copyright according to other files.

namespaceXP replied:
done

namespaceXP committed:
dbs-virtio-devices: add vsock packet.

Add VsockPacket, which provides a thin wrapper over the buffers used by
vsock.

Signed-off-by: Beiyue <beiyue@linux.alibaba.com>

ZizhengBian commented on crates/dbs-virtio-devices/Cargo.toml:
```diff
@@ -28,3 +30,4 @@ vm-memory = { version = "0.7.0", features = [ "backend-mmap", "backend-atomic" ]
 
 [features]
 virtio-mmio = []
+virtio-vsock = ["virtio-mmio"]
```
pls add a new line

ZizhengBian commented on crates/dbs-virtio-devices/src/vsock/backend/inner.rs:
```diff
@@ -0,0 +1,435 @@
+// Copyright 2022 Alibaba Cloud. All Rights Reserved.
```
the commit message tell us that you only add traits to vsock, but here is the implementation

ZizhengBian commented on crates/dbs-virtio-devices/src/vsock/backend/inner.rs:
```diff
@@ -0,0 +1,435 @@
+                        if self.stream_nonblocking.load(Ordering::SeqCst) {
+                            return Err(Error::from(ErrorKind::WouldBlock));
+                        } else {
+                            // - no read timeout: use channel 's `recv` function
```
an extra space in comment?

ZizhengBian commented on crates/dbs-virtio-devices/src/vsock/backend/mod.rs:
```diff
@@ -0,0 +1,67 @@
+#[derive(PartialEq, Eq, Hash, Debug, Clone)]
+pub enum VsockBackendType {
+    /// Unix Domain Socket
+    Uds,
```
Uds contains UnixStream and UnixDatagram, should we use more exact word

ZizhengBian commented on crates/dbs-virtio-devices/src/vsock/backend/mod.rs:
```diff
@@ -0,0 +1,67 @@
+    Test,
+}
+
+impl fmt::Display for VsockBackendType {
```
Why we need implement fmt::Display, Debug is not enough?

ZizhengBian commented on crates/dbs-virtio-devices/src/lib.rs:
```diff
@@ -119,6 +119,10 @@ pub enum Error {
+
+    #[cfg(feature = "virtio-vsock")]
+    #[error("virtio-vsock error: {0}")]
+    VirtioVsockError(#[from] self::vsock::VsockError),
```
Why add error in tcp backend's implementation patch

ZizhengBian commented on crates/dbs-virtio-devices/src/vsock/mod.rs:
```diff
@@ -6,3 +6,12 @@
 pub mod backend;
+
+#[derive(Debug, thiserror::Error)]
+pub enum VsockError {
```
Why add error in tcp backend's implementation patch

ZizhengBian commented on crates/dbs-virtio-devices/src/vsock/backend/uds.rs:
```diff
@@ -0,0 +1,92 @@
+
+impl VsockUnixBackend {
+    pub fn new(host_sock_path: String) -> Result<Self> {
+        info!("Open vsock uds: {}", host_sock_path);
```
should we use `info` log level

ZizhengBian commented on crates/dbs-virtio-devices/src/vsock/mod.rs:
```diff
@@ -6,12 +6,145 @@
+
+use self::packet::VsockPacket;
+
+mod defs {
```
Should we add defs in VsockPacket's patch

ZizhengBian commented on crates/dbs-virtio-devices/src/vsock/mod.rs:
```diff
@@ -6,12 +6,145 @@
+/// file descriptor and the set of events it's interested in. When such an event
+/// occurs, the control loop will route the event to the listener via
+/// `notify()`.
+pub trait VsockEpollListener: AsRawFd {
```
Should we add these traits in VsockPacket's patch

ZizhengBian commented on crates/dbs-virtio-devices/src/vsock/mod.rs:
```diff
@@ -6,12 +6,145 @@
 // found in the THIRD-PARTY file.
 
 pub mod backend;
+pub mod csm;
```
should we add csm mod in vsock packet's patch

ZizhengBian commented on crates/dbs-virtio-devices/src/vsock/muxer/mod.rs:
```diff
@@ -0,0 +1,25 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
```
Please add Alibaba's copyright

ZizhengBian commented on crates/dbs-virtio-devices/src/vsock/muxer/muxer_impl.rs:
```diff
@@ -0,0 +1,45 @@
+// Copyright 2022 Alibaba Cloud. All Rights Reserved.
```
Why this file is imcomplete?

ZizhengBian replied:
the comments and the content isn't match!

ZizhengBian commented on crates/dbs-virtio-devices/src/vsock/csm/connection.rs:
```diff
@@ -106,15 +106,15 @@ pub struct VsockConnection {
     /// The local CID. Most of the time this will be the constant `2` (the vsock
     /// host CID).
-    local_cid: u64,
+    pub(crate) local_cid: u64,
```
Why make them  pub(crate)

ZizhengBian commented on crates/dbs-virtio-devices/src/vsock/muxer/mod.rs:
```diff
@@ -13,8 +13,12 @@ pub mod muxer_rxq;
 
 use super::backend::{VsockBackend, VsockBackendType};
 use super::{VsockChannel, VsockEpollListener};
+pub use muxer_impl::VsockMuxer;
```
pub use self::muxer_impl::VsockMuxer?

ZizhengBian commented on crates/dbs-virtio-devices/src/vsock/muxer/muxer_impl.rs:
```diff
@@ -53,3 +67,799 @@ pub enum MuxerRx {
+    LocalStream(Box<dyn VsockStream>),
+}
+
+impl PartialEq for EpollListener {
```
Do we need partialeq?

ZizhengBian commented on crates/dbs-virtio-devices/src/epoll_handler.rs:
```diff
@@ -0,0 +1,310 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Is this file correct???

ZizhengBian commented on crates/dbs-virtio-devices/src/vsock/epoll_handler.rs:
```diff
@@ -0,0 +1,309 @@
+        }
+    }
+
+    pub(crate) fn handle_rxq_event(&mut self, mem: &AS::M) {
```
do we need pub(crate), same as below

ZizhengBian commented on crates/dbs-virtio-devices/src/vsock/device.rs:
```diff
@@ -0,0 +1,194 @@
+pub struct Vsock<AS: GuestAddressSpace, M: VsockGenericMuxer = VsockMuxer> {
+    cid: u64,
+    queue_sizes: Arc<Vec<u16>>,
+    pub(crate) device_info: VirtioDeviceInfo,
```
do we need pub(Crate)?

ZizhengBian commented on crates/dbs-virtio-devices/src/vsock/device.rs:
```diff
@@ -0,0 +1,194 @@
+}
+
+impl<AS: GuestAddressSpace, M: VsockGenericMuxer> Vsock<AS, M> {
+    pub(crate) fn new_with_muxer(
```
do we need pub(crate)?

ZizhengBian reviewed (changes requested):
need to distinguish patches more carefully

ZizhengBian commented on crates/dbs-virtio-devices/src/device.rs:
```diff
@@ -535,6 +538,16 @@ impl VirtioDeviceInfo {
 }
 
+impl PartialEq for VirtioDeviceInfo {
+    fn eq(&self, other: &VirtioDeviceInfo) -> bool {
```
Do we really need this?

ZizhengBian reviewed (changes requested):
need to distinguish patches more carefully

namespaceXP committed:
dbs-virtio-devices: add basic definitions for vsock.

Add basic definitions for vsock.

Signed-off-by: Beiyue <beiyue@linux.alibaba.com>

namespaceXP committed:
dbs-virtio-devices: add basic traits and impls for vsock backend.

Add basic traits and their implementations for vsock backend and vsock
connection.

Signed-off-by: Beiyue <beiyue@linux.alibaba.com>

namespaceXP committed:
dbs-virtio-devices: add connection state machine for vsock.

Add the connection state machine of vsock connection.

Signed-off-by: Beiyue <beiyue@linux.alibaba.com>

namespaceXP committed:
dbs-virtio-devices: add muxer for vsock.

Add VsockMuxer, the device-facing component of multiple vsock backends.

Signed-off-by: Beiyue <beiyue@linux.alibaba.com>

ZizhengBian commented on crates/dbs-virtio-devices/Cargo.toml:
```diff
@@ -28,3 +28,4 @@ vm-memory = { version = "0.7.0", features = [ "backend-mmap", "backend-atomic" ]
 
 [features]
 virtio-mmio = []
+virtio-vsock = ["virtio-mmio"]
```
add a blank line

ZizhengBian commented on crates/dbs-virtio-devices/src/vsock/backend/unix_stream.rs:
```diff
@@ -0,0 +1,92 @@
+}
+
+/// The backend implementation that using Unix Domain Socket.
+pub struct VsockUnixBackend {
```
VsockUnixStreamBackend?

ZizhengBian reviewed (commented):
Pls check sob in your commit log

ZizhengBian commented on crates/dbs-virtio-devices/Cargo.toml:
```diff
@@ -18,6 +18,7 @@ dbs-utils = { version = "0.1.0", path = "../dbs-utils" }
 epoll = { version = ">=4.0.1" }
 kvm-bindings = "0.5.0"
 kvm-ioctls = "0.11.0"
+libc = "0.2.39"
```
why add libc here？

ZizhengBian commented:
LGTM

namespaceXP committed:
dbs-utils: modify `test_rate_limiter_overconsumption` case.

Performance issue on test machine may led to inappropriate sleep time, so we
modify the test case to avoid test failure.

Signed-off-by: Beiyue <beiyue@linux.alibaba.com>
